### PR TITLE
Fixed wrong construction of LoHa weights, updated adapters conversion script

### PR DESCRIPTION
**Pull request by kovalexal**
This PR addresses the issue I found with the current LoHa adapter implementation. Right now there is a bug that does not take into account hada_t2 parameter during delta weights [calculation](https://github.com/huggingface/peft/blob/aaa7e9f44a6405af819e721d7ee7fc6dd190c980/src/peft/tuners/loha/layer.py#L160) both in train and inference.

This bug also exists in the original implementation, so I've created a similar [PR](https://github.com/KohakuBlueleaf/LyCORIS/pull/115) to fix it in LyCORIS repo.

I've included some tests that cover this functionality + fixed several errors in SD adapters conversion script (mentioned in [#978](https://github.com/huggingface/peft/pull/978)).

It would be great if this PR would get merged before the next release, so we won't need to provide some sort of backward compatibility for this.

[@BenjaminBossan](https://github.com/BenjaminBossan) your review is kindly appreciated.

### CODE REVIEW

File: examples/stable_diffusion/convert_sd_adapter_to_peft.py
Comment (BenjaminBossan) [inline examples/stable_diffusion/convert_sd_adapter_to_peft.py:291]: IIUC, once the bug is fixed, this code should be reverted to its previous state, is that right? How would we know if a user converts weights with vs without the bug?

Comment (kovalexal) [reply examples/stable_diffusion/convert_sd_adapter_to_peft.py:291]: > IIUC, once the bug is fixed, this code should be reverted to its previous state, is that right?

Yes, but it will depend on whether the provided checkpoint was trained with or without this fix. There is no obvious way to determine it right now...

We may add a flag to this conversion script, but it will be on the user's side to decide what should be done here.

Comment (BenjaminBossan) [reply examples/stable_diffusion/convert_sd_adapter_to_peft.py:291]: Yeah, I think we need to have an option to switch between before vs after the patch. If the patch is tied to a specific version, we could add an argument where users pass the version and we check if it is from before vs after the patch.

Comment (kovalexal) [reply examples/stable_diffusion/convert_sd_adapter_to_peft.py]: I added a CLI parameter to the conversion script to optionally enable/disable the fix through the weights.

> If the patch is tied to a specific version, we could add an argument where users pass the version and we check if it is from before vs after the patch.

As far as I see, current checkpoints out there do not contain information about LyCORIS version, so we probably have to manually force a user to provide, whether this fix should be applied or not (I also added a warning to highlight this problem).

File: examples/stable_diffusion/convert_sd_adapter_to_peft.py
Comment (BenjaminBossan) [inline examples/stable_diffusion/convert_sd_adapter_to_peft.py:209]: I think the message should be expanded a little bit to make it very clear to the user when to use this flag: If the weights they use are created before the mentioned PR was merged, this argument should be True, if they're from after the PR was merged, it should be False.

File: examples/stable_diffusion/convert_sd_adapter_to_peft.py
Comment (BenjaminBossan) [inline examples/stable_diffusion/convert_sd_adapter_to_peft.py:291]: Let's not use word as "Right now", because they will be outdated quickly. Maybe something along the lines of: "This code block accommodates a bug that existed for some time that resulted in accidentally using hada_t1 weight for hada_t2, see ...".

WDYT?

File: examples/stable_diffusion/convert_sd_adapter_to_peft.py
Comment (BenjaminBossan) [inline examples/stable_diffusion/convert_sd_adapter_to_peft.py:323]: Again, let's avoid words like "Current".

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the update. I made some suggestions regarding the wording, to minimize the possibility for confusion. Other than that, we should wait for the bugfix in the LyCORIS repo to be merged, right?

File: examples/stable_diffusion/convert_sd_adapter_to_peft.py
Comment (kovalexal) [reply examples/stable_diffusion/convert_sd_adapter_to_peft.py]: Okay, I replaced it with LyCORIS library version description.

**Comment by kovalexal** [comment]
> Other than that, we should wait for the bugfix in the LyCORIS repo to be merged, right?

Yep, the PR in LyCORIS repo got merged, so we can proceed with this PR.

**Comment by kovalexal** [comment]
@BenjaminBossan Sorry to disturb you, but are there any open questions for this PR that need to be addressed by me?

**Review by BenjaminBossan — APPROVED** [review]
Fantastic, thanks for addressing the last few comments. Let me check with my colleagues if we want another pair of eyes to take a look, otherwise I'll merge later.

**Pull request merged**
